File: backend/src/agents/enrichmentAgent.ts
import { extractContent, validateUrl } from "../tools/contentExtractor";
import { analyzeContentWithTrace, type AnalysisTrace } from "../chains/analysisChain";
import { suggestTags } from "../chains/taggingChain";
import { evaluateSummaryQualityWithTrace, type JudgeTrace } from "../chains/judgeChain";
import { getEmbedderAgent } from "./embedderAgent";
import type {
  EnrichmentOptions,
  EnrichmentResult,
  EnrichmentError,
  DetectedContentType,
  ContentTypeClassification,
  AnalyzerContext,
} from "../types/schemas";
import type { AgentTrace } from "../services/jobStorage";

// PHASE 1: Content-Type Routing
import { ContentTypeClassifierAgent } from "./ContentTypeClassifierAgent";
import { BaseAnalyzerAgent } from "./analyzers/BaseAnalyzerAgent";
import { ArticleAnalyzerAgent } from "./analyzers/ArticleAnalyzerAgent";
import { PaperAnalyzerAgent } from "./analyzers/PaperAnalyzerAgent";
import { VideoAnalyzerAgent } from "./analyzers/VideoAnalyzerAgent";
import { SocialAnalyzerAgent } from "./analyzers/SocialAnalyzerAgent";
import { DocumentAnalyzerAgent } from "./analyzers/DocumentAnalyzerAgent";
import { GenericAnalyzerAgent } from "./analyzers/GenericAnalyzerAgent";

/**
 * Enrichment Agent - Main orchestrator for bookmark enrichment
 *
 * Coordinates the complete enrichment pipeline:
 * 1. Extract content from URL (Readability + axios)
 * 2. Analyze content (LangChain + GPT-4o-mini)
 * 3. Suggest tags (LangChain + GPT-4o-mini)
 *
 * Features:
 * - Sequential processing with error tracking
 * - Graceful degradation (returns partial results on failure)
 * - Processing time tracking
 * - Detailed error reporting
 */

interface EnrichmentProgress {
  step: "extraction" | "classification" | "analysis" | "tagging" | "embedding" | "completed";
  message: string;
  timestamp: Date;
}

export class EnrichmentAgent {
  private errors: EnrichmentError[] = [];
  private progressCallbacks: Array<(progress: EnrichmentProgress) => void> = [];
  private agentTraces: AgentTrace[] = []; // Collect detailed LLM traces

  /**
   * Register a callback to track enrichment progress
   */
  onProgress(callback: (progress: EnrichmentProgress) => void) {
    this.progressCallbacks.push(callback);
  }

  /**
   * Emit progress update
   */
  private emitProgress(step: EnrichmentProgress["step"], message: string) {
    const progress: EnrichmentProgress = {
      step,
      message,
      timestamp: new Date(),
    };

    this.progressCallbacks.forEach((callback) => callback(progress));
  }

  /**
   * Record an error with recovery status
   */
  private recordError(
    step: "extraction" | "classification" | "analysis" | "tagging" | "embedding",
    error: unknown,
    recoverable: boolean = false
  ) {
    const errorMessage =
      error instanceof Error ? error.message : String(error);

    this.errors.push({
      step,
      error: errorMessage,
      timestamp: new Date(),
      recoverable,
    });

    console.error(`[EnrichmentAgent] Error in ${step}:`, errorMessage);
  }

  /**
   * Enrich a bookmark from its URL
   *
   * @param options - Enrichment options including URL and existing tags
   * @returns Complete enrichment result with metadata
   */
  async enrich(options: EnrichmentOptions): Promise<EnrichmentResult> {
    const startTime = Date.now();
    this.errors = []; // Reset errors for new enrichment
    this.agentTraces = []; // Reset traces for new enrichment

    // console.log(`\n[EnrichmentAgent] Starting enrichment for: ${options.url}`);

    // Step 1: Validate URL
    this.emitProgress("extraction", "Validating URL...");
    const isValid = await validateUrl(options.url);
    if (!isValid) {
      throw new Error(`The URL could not be accessed. Please check that the link is correct and the website is available.`);
    }

    // Step 2: Extract content
    this.emitProgress("extraction", "Extracting content...");
    let extractedContent;
    try {
      extractedContent = await extractContent(options.url);
      // console.log(`[EnrichmentAgent] Extracted ${extractedContent.cleanText.length} characters`);
    } catch (error) {
      this.recordError("extraction", error, false);
      throw new Error(`Failed to extract content: ${error}`);
    }

    // If extraction confidence is too low, warn but continue
    if (extractedContent.extractionConfidence < 0.5) {
      console.warn(
        "[EnrichmentAgent] Low extraction confidence:",
        extractedContent.extractionConfidence
      );
    }

    // PHASE 1: Step 2.5: Classify content type
    this.emitProgress("classification", "Classifying content type...");
    let contentTypeClassification: ContentTypeClassification;

    try {
      const classifier = new ContentTypeClassifierAgent();
      const classificationStartTime = new Date();

      contentTypeClassification = await classifier.classify(
        options.url,
        extractedContent
      );

      // Store classification trace
      this.agentTraces.push({
        agentName: "ContentTypeClassifier",
        startTime: classificationStartTime,
        endTime: new Date(),
        duration: Date.now() - classificationStartTime.getTime(),
        input: {
          url: options.url,
          domain: extractedContent.domain,
          contentLength: extractedContent.cleanText.length,
        },
        output: {
          type: contentTypeClassification.type,
          confidence: contentTypeClassification.confidence,
          method: contentTypeClassification.method,
        },
        metadata: {
          indicators: contentTypeClassification.indicators,
        },
      });

      console.log(
        `[EnrichmentAgent] Classified as '${contentTypeClassification.type}' ` +
          `(confidence: ${contentTypeClassification.confidence.toFixed(2)}, ` +
          `method: ${contentTypeClassification.method})`
      );
    } catch (error) {
      this.recordError("classification", error, true);
      // Graceful degradation: use 'other' type
      contentTypeClassification = {
        type: "other",
        confidence: 0.5,
        method: "heuristic",
      };
      console.warn(
        "[EnrichmentAgent] Classification failed, using fallback type 'other'"
      );
    }

    // Step 3: Analyze content with specialized analyzer + Quality Gates (3-layer system)
    this.emitProgress(
      "analysis",
      `Analyzing ${contentTypeClassification.type} content with AI...`
    );
    let analysis: EnhancedAnalysisResult;
    let qualityGateStatus: "passed" | "needs_review" = "passed";

    // Load retry configuration from environment
    const maxRetries = parseInt(process.env.ENRICHMENT_MAX_RETRIES || "2", 10);
    const retryDelayMs = parseInt(
      process.env.ENRICHMENT_RETRY_DELAY_MS || "2000",
      10
    );
    const enablePreValidation =
      process.env.ENRICHMENT_ENABLE_PRE_VALIDATION !== "false";
    const enableJudge = process.env.ENRICHMENT_ENABLE_JUDGE !== "false";

    console.log(
      `[EnrichmentAgent] Quality gates: pre-validation=${enablePreValidation}, judge=${enableJudge}, maxRetries=${maxRetries}`
    );

    // Retry loop for analysis + quality gates
    let retryCount = 0;
    let analysisSucceeded = false;

    while (retryCount <= maxRetries && !analysisSucceeded) {
      try {
        // Generate analysis (with or without LLM)
        if (options.skipAnalysis) {
          console.log("[EnrichmentAgent] Skipping analysis (option set)");
          analysis = {
            title: options.userTitle || extractedContent.title,
            summary:
              options.userSummary ||
              options.userNotes ||
              "No summary available",
            tags: options.userTags || [],
            keyPoints: [],
            tone: "unknown",
            contentMetrics: this.calculateBasicMetrics(
              extractedContent.cleanText
            ),
            confidence: 0.5,
            modelUsed: "skipped",
          };
          analysisSucceeded = true; // Skip quality gates if analysis is skipped
          break;
        } else {
          // PHASE 1: Select specialized analyzer based on content type
          const analyzer = this.selectAnalyzer(
            contentTypeClassification.type
          );

          const analyzerContext: AnalyzerContext = {
            extractedContent,
            contentTypeClassification,
            userContext: {
              userTitle: options.userTitle,
              userSummary: options.userSummary,
              userTags: options.userTags,
            },
          };

          const analysisStartTime = new Date();
          const { result: analysisResult, trace: analysisTrace } =
            await analyzer.analyze(analyzerContext);
          analysis = analysisResult;

          // Store analyzer trace
          this.agentTraces.push({
            agentName: `${this.getAnalyzerName(contentTypeClassification.type)}Analyzer`,
            startTime: analysisStartTime,
            endTime: new Date(),
            duration: analysisTrace.duration,
            input: {
              extractedTitle: extractedContent.title,
              contentLength: extractedContent.cleanText.length,
              contentType: extractedContent.contentType,
              detectedContentType: contentTypeClassification.type,
              userContext: analyzerContext.userContext,
            },
            output: {
              title: analysisResult.title,
              summaryLength: analysisResult.summary.length,
              tags: analysisResult.tags,
              keyPointsCount: analysisResult.keyPoints.length,
              tone: analysisResult.tone,
              confidence: analysisResult.confidence,
            },
            llmTrace: analysisTrace,
            metadata: retryCount > 0 ? { retryAttempt: retryCount } : undefined,
          });

          console.log(
            `[EnrichmentAgent] Generated ${contentTypeClassification.type} analysis (attempt ${retryCount + 1}/${maxRetries + 1}): ${analysis.title}`
          );
        }

        // LAYER 1: Pre-validation (fast, deterministic, $0 cost)
        if (enablePreValidation && !options.skipAnalysis) {
          const preValidation = this.validateEnrichmentQuality(
            analysis,
            extractedContent
          );

          if (!preValidation.valid) {
            this.recordError(
              "analysis",
              new Error(
                preValidation.reason || "Pre-validation failed"
              ),
              true
            );

            if (retryCount < maxRetries) {
              retryCount++;
              console.log(
                `[EnrichmentAgent] Pre-validation failed: ${preValidation.reason}. Retry ${retryCount}/${maxRetries}...`
              );
              await new Promise((resolve) =>
                setTimeout(resolve, retryDelayMs)
              );
              continue; // Retry analysis
            } else {
              // Max retries exceeded
              console.warn(
                `[EnrichmentAgent] Max retries exceeded after pre-validation failures`
              );
              qualityGateStatus = "needs_review";
              break; // Exit retry loop
            }
          } else {
            console.log(`[EnrichmentAgent] Pre-validation passed`);
          }
        }

        // LAYER 2: Conditional Judge (LLM-based quality evaluation)
        const shouldJudge =
          enableJudge &&
          !options.skipAnalysis &&
          (analysis.confidence <= 0.5 ||
            extractedContent.cleanText.length > 5000 ||
            contentTypeClassification.type === "pdf" ||
            contentTypeClassification.type === "video");

        if (shouldJudge) {
          this.emitProgress("analysis", "Evaluating summary quality...");

          try {
            const judgeStartTime = new Date();
            const { result: qualityCheck, trace: judgeTrace } =
              await evaluateSummaryQualityWithTrace(
                analysis,
                extractedContent
              );

            // Store judge trace
            this.agentTraces.push({
              agentName: "Judge",
              startTime: judgeStartTime,
              endTime: new Date(),
              duration: judgeTrace.duration,
              input: {
                summary: analysis.summary.substring(0, 200) + "...",
                sourceContentLength: extractedContent.cleanText.length,
              },
              output: {
                verdict: qualityCheck.overall_verdict,
                accuracy: qualityCheck.accuracy,
                comprehensiveness: qualityCheck.comprehensiveness,
                formatting: qualityCheck.formatting,
                completeness: qualityCheck.completeness,
                issues: qualityCheck.issues,
              },
              llmTrace: judgeTrace,
              metadata: retryCount > 0 ? { retryAttempt: retryCount } : undefined,
            });

            if (qualityCheck.overall_verdict === "fail") {
              this.recordError(
                "analysis",
                new Error(
                  `Judge rejected: ${qualityCheck.issues.join(", ")}`
                ),
                true
              );

              if (retryCount < maxRetries) {
                retryCount++;
                console.log(
                  `[EnrichmentAgent] Judge rejected (${qualityCheck.issues.join(", ")}). Retry ${retryCount}/${maxRetries}...`
                );
                await new Promise((resolve) =>
                  setTimeout(resolve, retryDelayMs)
                );
                continue; // Retry analysis
              } else {
                // Max retries exceeded
                console.warn(
                  `[EnrichmentAgent] Max retries exceeded after judge rejections`
                );
                qualityGateStatus = "needs_review";
                break; // Exit retry loop
              }
            } else {
              console.log(
                `[EnrichmentAgent] Judge approved: ${qualityCheck.reasoning}`
              );
            }
          } catch (error) {
            this.recordError("analysis", error, true);
            console.warn(
              "[EnrichmentAgent] Judge evaluation failed, continuing with current analysis"
            );
          }
        } else {
          const reason = !enableJudge
            ? "judge disabled"
            : analysis.confidence > 0.5
            ? "high confidence"
            : "short content";
          console.log(`[EnrichmentAgent] Skipping judge: ${reason}`);
        }

        // All quality gates passed
        analysisSucceeded = true;
        console.log(
          `[EnrichmentAgent] All quality gates passed for analysis`
        );
      } catch (error) {
        this.recordError("analysis", error, true);
        console.error(
          `[EnrichmentAgent] Analysis error (attempt ${retryCount + 1}/${maxRetries + 1}):`,
          error
        );

        if (retryCount < maxRetries) {
          retryCount++;
          console.log(
            `[EnrichmentAgent] Retrying analysis (${retryCount}/${maxRetries})...`
          );
          await new Promise((resolve) => setTimeout(resolve, retryDelayMs));
          continue;
        } else {
          // Max retries exceeded - use graceful degradation
          console.warn(
            `[EnrichmentAgent] Max retries exceeded, using fallback analysis`
          );
          analysis = {
            title: options.userTitle || extractedContent.title || "Untitled",
            summary:
              options.userSummary ||
              `Content from ${extractedContent.domain}: ${extractedContent.title}. AI analysis failed - manual review needed.`,
            tags: options.userTags || [
              contentTypeClassification.type,
              "needs-review",
            ],
            keyPoints: ["Analysis failed - requires manual review"],
            tone: "unknown",
            contentMetrics: this.calculateBasicMetrics(
              extractedContent.cleanText
            ),
            confidence: 0.2,
            modelUsed: "fallback",
          };
          qualityGateStatus = "needs_review";
          break; // Exit retry loop
        }
      }
    }

    // Step 4: Use tags from analysis (no separate tagging chain needed)
    // OPTIMIZATION: Eliminated redundant tagging chain - analysis already generates high-quality tags
    this.emitProgress("tagging", "Organizing tags...");
    let tagging;

    // Use tags from analysis only (saves ~1,600 tokens per enrichment)
    tagging = {
      tags: analysis.tags?.length > 0
        ? analysis.tags.slice(0, 5) // Limit to 5 tags maximum
        : [extractedContent.contentType, extractedContent.domain] // Fallback tags
    };

    // console.log(`[EnrichmentAgent] Using tags from analysis (${tagging.tags.length}): ${tagging.tags.join(', ')}`);

    // Step 5: Generate embeddings (multiple types)
    this.emitProgress("embedding", "Generating vector embeddings...");
    let embedding: number[] | undefined;
    let summaryEmbedding: number[] | undefined;
    let fullContent: { chunks?: number[][]; text?: string } | undefined;
    let embeddedAt: Date | undefined;

    try {
      if (options.skipEmbedding) {
        console.log("[EnrichmentAgent] Skipping embedding (option set)");
      } else {
        const embedder = getEmbedderAgent();

        // Generate multiple embeddings
        // Note: Tags are intentionally excluded from embeddings
        // Semantic relationships are now handled by concepts and entities in the graph pipeline
        const embeddings = await embedder.embedMulti({
          title: analysis.title,
          summary: analysis.summary,
          fullContent: extractedContent.cleanText.length > 10000
            ? extractedContent.cleanText
            : undefined,
          useCache: true,
        });

        embedding = embeddings.combined;
        summaryEmbedding = embeddings.summary;

        if (embeddings.chunks) {
          fullContent = {
            text: extractedContent.cleanText,
            chunks: embeddings.chunks,
          };
        }

        embeddedAt = new Date();

        console.log(
          `[EnrichmentAgent] Generated ${embeddings.chunks ? '3' : '2'} embedding types`
        );
      }
    } catch (error) {
      this.recordError("embedding", error, true);
      // Graceful degradation: bookmark is still usable without embedding
      console.warn(
        "[EnrichmentAgent] Failed to generate embeddings, continuing without them"
      );
    }

    // Step 6: Compile final result
    this.emitProgress("completed", "Enrichment complete!");
    const processingTime = Date.now() - startTime;

    const result: EnrichmentResult & { qualityGateStatus: "passed" | "needs_review" } = {
      url: options.url,
      title: analysis.title, // Use improved title from analysis (not raw extracted title)
      domain: extractedContent.domain,
      contentType: extractedContent.contentType,
      extractedContent: {
        rawText: extractedContent.rawText,
        cleanText: extractedContent.cleanText,
        images: extractedContent.images,
        metadata: extractedContent.metadata,
      },
      analysis,
      tagging,
      embedding,
      summaryEmbedding,
      fullContent,
      embeddedAt,
      enrichedAt: new Date(),
      modelUsed: process.env.AI_MODEL || "gpt-4o-mini-2024-07-18",
      processingTimeMs: processingTime,
      qualityGateStatus, // Add quality gate status to result
    };

    console.log(
      `[EnrichmentAgent] Completed in ${processingTime}ms with status: ${qualityGateStatus}${this.errors.length > 0 ? ` (${this.errors.length} error(s))` : ""}`
    );

    return result;
  }

  // ========================================================================
  // PHASE 1: Analyzer Factory & Helper Methods
  // ========================================================================

  /**
   * Select specialized analyzer based on detected content type
   */
  private selectAnalyzer(type: DetectedContentType): BaseAnalyzerAgent {
    switch (type) {
      case "article":
        return new ArticleAnalyzerAgent();
      case "paper":
        return new PaperAnalyzerAgent();
      case "video":
        return new VideoAnalyzerAgent();
      case "social":
        return new SocialAnalyzerAgent();
      case "document":
        return new DocumentAnalyzerAgent();
      default:
        return new GenericAnalyzerAgent();
    }
  }

  /**
   * Get human-readable analyzer name for tracing
   */
  private getAnalyzerName(type: DetectedContentType): string {
    switch (type) {
      case "article":
        return "Article";
      case "paper":
        return "Paper";
      case "video":
        return "Video";
      case "social":
        return "Social";
      case "document":
        return "Document";
      default:
        return "Generic";
    }
  }

  /**
   * Calculate basic content metrics (used for fallback)
   */
  private calculateBasicMetrics(text: string): {
    readingLevel: number;
    wordCount: number;
    estimatedReadTime: number;
  } {
    const wordCount = text
      .split(/\s+/)
      .filter((word) => word.length > 0).length;
    return {
      readingLevel: 10.0, // Default grade level
      wordCount,
      estimatedReadTime: Math.max(1, Math.ceil(wordCount / 200)), // 200 words/min
    };
  }

  /**
   * Pre-validation: Fast, deterministic check for fallback content
   * Returns true if content should be rejected (needs retry/review)
   * Cost: $0 (no LLM calls)
   */
  private validateEnrichmentQuality(
    analysis: EnhancedAnalysisResult,
    extractedContent: ExtractedContent
  ): { valid: boolean; reason?: string } {
    // 1. Check for explicit fallback markers
    const fallbackPatterns = [
      /AI analysis failed/i,
      /manual review needed/i,
      /requires manual review/i,
      /analysis failed/i,
      /could not analyze/i,
      /failed to extract/i,
      /error extracting/i,
    ];

    for (const pattern of fallbackPatterns) {
      if (pattern.test(analysis.summary)) {
        return {
          valid: false,
          reason: `Fallback pattern detected in summary: ${pattern}`,
        };
      }
    }

    // 2. Check for suspiciously short summaries (< 50 chars)
    if (
      analysis.summary.length < 50 &&
      extractedContent.cleanText.length > 500
    ) {
      return {
        valid: false,
        reason: `Summary too short (${analysis.summary.length} chars) for content length (${extractedContent.cleanText.length} chars)`,
      };
    }

    // 3. Check for missing critical fields
    if (!analysis.title || analysis.title === "Untitled") {
      return {
        valid: false,
        reason: "Missing or default title",
      };
    }

    // 4. Check for low confidence with fallback model
    if (analysis.confidence <= 0.3 && analysis.modelUsed === "fallback") {
      return {
        valid: false,
        reason: "Low confidence with fallback model",
      };
    }

    // 5. Check for generic error tags
    const errorTags = ["needs-review", "error", "failed", "fallback"];
    const hasErrorTag = analysis.tags.some((tag) =>
      errorTags.includes(tag.toLowerCase())
    );
    if (hasErrorTag && analysis.confidence < 0.5) {
      return {
        valid: false,
        reason: `Error tag present with low confidence: ${analysis.tags.join(", ")}`,
      };
    }

    return { valid: true };
  }

  // ========================================================================

  /**
   * Get errors that occurred during enrichment
   */
  getErrors(): EnrichmentError[] {
    return this.errors;
  }

  /**
   * Check if enrichment had any errors
   */
  hasErrors(): boolean {
    return this.errors.length > 0;
  }

  /**
   * Check if enrichment had non-recoverable errors
   */
  hasCriticalErrors(): boolean {
    return this.errors.some((error) => !error.recoverable);
  }

  /**
   * Get detailed agent traces with LLM observability data
   */
  getAgentTraces(): AgentTrace[] {
    return this.agentTraces;
  }
}

/**
 * Convenience function to enrich a single URL
 */
export async function enrichUrl(
  url: string,
  existingTags: string[] = []
): Promise<EnrichmentResult> {
  const agent = new EnrichmentAgent();

  // Optional: Log progress
  // agent.onProgress((progress) => {
  //   console.log(`[Progress] ${progress.step}: ${progress.message}`);
  // });

  return agent.enrich({
    url,
    existingTags,
  });
}

File: ai-service/agents/enrichment_agent.py
"""
EnrichmentAgent - Main orchestrator for the enrichment pipeline.

Coordinates content classification, analysis, and embedding generation.
Python equivalent of backend/src/workers/enrichmentWorker.ts logic.
"""
from typing import List, Optional, Callable, Awaitable, Any
import time
from datetime import datetime

from langchain_openai import OpenAIEmbeddings

from agents.content_type_classifier import ContentTypeClassifierAgent
from agents.analyzers.article_analyzer import ArticleAnalyzerAgent
from agents.analyzers.paper_analyzer import PaperAnalyzerAgent
from agents.analyzers.video_analyzer import VideoAnalyzerAgent
from agents.analyzers.social_analyzer import SocialAnalyzerAgent
from agents.analyzers.document_analyzer import DocumentAnalyzerAgent
from agents.analyzers.generic_analyzer import GenericAnalyzerAgent
from schemas.enrichment import (
    ExtractedContent,
    DetectedContentType,
    AnalyzerContext,
    EnhancedAnalysisResult,
    AnalysisTrace,
    EnrichmentResult,
)
from utils.cost_calculator import calculate_embedding_cost, estimate_tokens
from config import settings


class EnrichmentAgent:
    """
    Main orchestrator for bookmark enrichment pipeline.

    Pipeline stages:
    1. Content classification (heuristic + LLM fallback)
    2. Specialized analysis (content-type specific)
    3. Embedding generation (OpenAI text-embedding-3-small)

    Returns complete enrichment result with traces for observability.
    """

    def __init__(self):
        """Initialize enrichment agent."""
        self.classifier = ContentTypeClassifierAgent()
        self.progress_callback: Optional[Callable[[int], Awaitable[None]]] = None
        self.agent_traces: List[AnalysisTrace] = []

    def on_progress(self, callback: Callable[[int], Awaitable[None]]):
        """
        Register progress callback for real-time updates.

        Args:
            callback: Async function that receives progress percentage (0-100)
        """
        self.progress_callback = callback

    async def enrich(
        self,
        bookmark_id: str,
        url: str,
        extracted_content: ExtractedContent,
        user_tags: List[str],
        user_id: str
    ) -> EnrichmentResult:
        """
        Execute full enrichment pipeline.

        Args:
            bookmark_id: Bookmark ID
            url: Content URL
            extracted_content: Extracted content from URL
            user_tags: User-provided tags
            user_id: User ID

        Returns:
            Complete enrichment result with analysis and embedding
        """
        pipeline_start = time.time()
        self.agent_traces = []

        try:
            # Step 1: Classify content type (10% progress)
            await self._update_progress(10)
            classification = await self._classify_content(url, extracted_content)

            # Step 2: Analyze with specialized analyzer (50% progress)
            await self._update_progress(50)
            analysis, analysis_trace = await self._analyze_content(
                extracted_content,
                classification.type,
                user_tags
            )
            self.agent_traces.append(analysis_trace)

            # Step 3: Generate embedding (80% progress)
            await self._update_progress(80)
            embedding, embedding_trace = await self._generate_embedding(
                analysis,
                user_tags
            )
            self.agent_traces.append(embedding_trace)

            # Complete (100% progress)
            await self._update_progress(100)

            # Calculate total cost and latency
            total_cost = sum(trace.cost_usd or 0.0 for trace in self.agent_traces)
            total_latency = int((time.time() - pipeline_start) * 1000)

            return EnrichmentResult(
                bookmark_id=bookmark_id,
                url=url,
                title=analysis.title,
                summary=analysis.summary,
                tags=analysis.tags,
                key_points=analysis.key_points,
                tone=analysis.tone,
                content_type=classification.type,
                content_metrics=analysis.content_metrics,
                embedding=embedding,
                confidence=analysis.confidence,
                analysis_traces=self.agent_traces,
                total_cost_usd=total_cost,
                total_latency_ms=total_latency,
                enriched_at=datetime.now()
            )

        except Exception as error:
            print(f"[EnrichmentAgent] Pipeline failed: {error}")
            raise

    async def _classify_content(
        self,
        url: str,
        extracted_content: ExtractedContent
    ) -> Any:
        """
        Step 1: Classify content type.

        Args:
            url: Content URL
            extracted_content: Extracted content

        Returns:
            ContentTypeClassification
        """
        start_time = time.time()

        classification = await self.classifier.classify(url, extracted_content)

        latency_ms = int((time.time() - start_time) * 1000)

        # Create trace for classification
        trace = AnalysisTrace(
            step="classification",
            agent="ContentTypeClassifierAgent",
            model="gpt-4o-mini-2024-07-18" if classification.method == "llm" else None,
            input_tokens=None,
            output_tokens=None,
            cost_usd=0.0003 if classification.method == "llm" else 0.0,  # Estimated
            latency_ms=latency_ms,
            success=True,
            metadata={
                "content_type": classification.type.value,
                "confidence": classification.confidence,
                "method": classification.method,
                "indicators": classification.indicators
            }
        )
        self.agent_traces.append(trace)

        return classification

    async def _analyze_content(
        self,
        extracted_content: ExtractedContent,
        content_type: DetectedContentType,
        user_tags: List[str]
    ) -> tuple[EnhancedAnalysisResult, AnalysisTrace]:
        """
        Step 2: Analyze content with specialized analyzer.

        Args:
            extracted_content: Extracted content
            content_type: Detected content type
            user_tags: User-provided tags

        Returns:
            Tuple of (analysis result, trace)
        """
        # Select appropriate analyzer
        analyzer = self._select_analyzer(content_type)

        # Build analyzer context
        context = AnalyzerContext(
            url=extracted_content.url,
            title=extracted_content.title,
            clean_text=extracted_content.clean_text,
            content_type=content_type,
            user_provided_title=None,  # Could be added in future
            user_provided_summary=None,  # Could be added in future
            user_tags=user_tags,
            metadata=extracted_content.metadata
        )

        # Run analysis
        result, trace = await analyzer.analyze(context)

        return result, trace

    def _select_analyzer(self, content_type: DetectedContentType):
        """
        Factory method to select appropriate analyzer.

        Args:
            content_type: Detected content type

        Returns:
            Analyzer instance
        """
        if content_type == DetectedContentType.ARTICLE:
            return ArticleAnalyzerAgent()
        elif content_type == DetectedContentType.PAPER:
            return PaperAnalyzerAgent()
        elif content_type == DetectedContentType.VIDEO:
            return VideoAnalyzerAgent()
        elif content_type == DetectedContentType.SOCIAL:
            return SocialAnalyzerAgent()
        elif content_type == DetectedContentType.DOCUMENT:
            return DocumentAnalyzerAgent()
        else:  # DetectedContentType.OTHER
            return GenericAnalyzerAgent()

    async def _generate_embedding(
        self,
        analysis: EnhancedAnalysisResult,
        user_tags: List[str]
    ) -> tuple[List[float], AnalysisTrace]:
        """
        Step 3: Generate embedding vector.

        Args:
            analysis: Analysis result
            user_tags: User-provided tags

        Returns:
            Tuple of (embedding vector, trace)
        """
        start_time = time.time()

        # Combine title + summary + tags for embedding
        all_tags = list(set(analysis.tags + user_tags))
        embedding_text = f"{analysis.title}\n\n{analysis.summary}\n\nTags: {', '.join(all_tags)}"

        try:
            # Generate embedding using OpenAI
            embeddings = OpenAIEmbeddings(
                model=settings.EMBEDDING_MODEL,
                openai_api_key=settings.OPENAI_API_KEY
            )

            embedding_vector = await embeddings.aembed_query(embedding_text)

            latency_ms = int((time.time() - start_time) * 1000)

            # Estimate tokens and cost
            tokens_used = estimate_tokens(embedding_text)
            cost_usd = calculate_embedding_cost(settings.EMBEDDING_MODEL, tokens_used)

            trace = AnalysisTrace(
                step="embedding",
                agent="OpenAIEmbeddings",
                model=settings.EMBEDDING_MODEL,
                input_tokens=tokens_used,
                output_tokens=0,
                cost_usd=cost_usd,
                latency_ms=latency_ms,
                success=True,
                metadata={
                    "dimensions": len(embedding_vector),
                    "text_length": len(embedding_text)
                }
            )

            return embedding_vector, trace

        except Exception as error:
            print(f"[EnrichmentAgent] Embedding generation failed: {error}")

            # Return zero vector as fallback
            latency_ms = int((time.time() - start_time) * 1000)
            trace = AnalysisTrace(
                step="embedding",
                agent="OpenAIEmbeddings",
                model=settings.EMBEDDING_MODEL,
                input_tokens=0,
                output_tokens=0,
                cost_usd=0.0,
                latency_ms=latency_ms,
                success=False,
                error_message=str(error)
            )

            return [0.0] * 1536, trace  # Return zero vector

    async def _update_progress(self, percentage: int):
        """
        Update progress via callback if registered.

        Args:
            percentage: Progress percentage (0-100)
        """
        if self.progress_callback:
            await self.progress_callback(percentage)
